src/Models/ArduinoDeviceBase.ts

```diff
@@ -232,41 +233,28 @@ export abstract class ArduinoDeviceBase implements Device {
     }
   }
 
-  async generateCrc(
-      context: vscode.ExtensionContext, channel: vscode.OutputChannel) {
+  async generateCrc(channel: vscode.OutputChannel): Promise<void> {
     if (!(vscode.workspace.workspaceFolders &&
           vscode.workspace.workspaceFolders.length > 0)) {
-      const message = 'No workspace opened.';
-      vscode.window.showWarningMessage(message);
-      utils.channelShowAndAppendLine(channel, message);
-      return false;
+      throw new Error('No workspace opened.');

```

Comment:
can we define a new class for this kind of error? There are several errors related to environment like workspace not opened, or device path not found. also we may need to provide suggestions to users how they can fix it.

Reply:
fixed.